File: Sheet 7 (STLs I)/G.cpp
#include <iostream>
#include <deque>
#include <string>
using namespace std;
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    bool reverse=false;
    int n,tmp;
    cin >> n;
    deque<int> D;
    while (n--){
        string a;
        cin >> a;
        if (a=="toFront"){
            cin >> tmp; 
            if (!reverse) D.push_front(tmp);
            else D.push_back(tmp);
        }
        else if (a=="push_back"){
            cin >> tmp; 
            if (!reverse) D.push_back(tmp);
            else D.push_front(tmp);
        }
        else if (a=="reverse"){reverse=!reverse;}
        else if(a=="front"){
            if (!D.empty() && !reverse) {cout << D.front() << "\n"; D.pop_front();}
            else if (!D.empty() && reverse) {cout << D.back() << "\n"; D.pop_back();}
            else cout << "No job for Ada?\n";
        }
        else if(a=="back"){
            if (!D.empty() && !reverse){cout << D.back() << "\n"; D.pop_back();}
            else if (!D.empty() && reverse) {cout << D.front() << "\n"; D.pop_front();}
            else cout << "No job for Ada?\n";
        }
    }
}

File: Sheet 7 (STLs I)/G.py
from collections import deque

num_queries = int(input())  
queue = deque()
flip = False  
output = []  

for _ in range(num_queries):
    command = input().split()

    if command[0] == "toFront":
        value = int(command[1])
        if flip:
            queue.append(value)  
        else:
            queue.appendleft(value)  

    elif command[0] == "push_back":
        value = int(command[1])
        if flip:
            queue.appendleft(value)  
        else:
            queue.append(value)  

    elif command[0] == "front":
        if not queue:
            output.append("No job for Ada?")
        else:
            if flip:
                output.append(str(queue.pop()))  
            else:
                output.append(str(queue.popleft()))  

    elif command[0] == "back":
        if not queue:
            output.append("No job for Ada?")
        else:
            if flip:
                output.append(str(queue.popleft()))  
            else:
                output.append(str(queue.pop()))  

    elif command[0] == "reverse":
        flip = not flip  

print("\n".join(output))
